src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -113,111 +112,81 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
+                    var oldProjectStateChecksums = await solution.SolutionState
+                        .GetRequiredProjectState(projectId)
+                        .GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
 
-                // added project
-                foreach (var (projectId, newProjectChecksums) in newMap)
-                {
-                    if (!oldMap.ContainsKey(projectId))
-                    {
-                        var projectInfo = await _assetProvider.CreateProjectInfoAsync(projectId, newProjectChecksums.Checksum, cancellationToken).ConfigureAwait(false);
-                        solution = solution.AddProject(projectInfo);
-                    }
-                }
+                    var newProjectStateChecksums = await _assetProvider.GetAssetAsync<ProjectStateChecksums>(
+                        assetHint: projectId, newProjectChecksum, cancellationToken).ConfigureAwait(false);
 
-                // remove all project references from projects that changed. this ensures exceptions will not occur for
-                // cyclic references during an incremental update.
-                foreach (var (projectId, newProjectChecksums) in newMap)
-                {
-                    if (!oldMap.TryGetValue(projectId, out var oldProjectChecksums))
-                    {
+                    // Ignore if they didn't actually change.
+                    if (oldProjectStateChecksums.Checksum == newProjectStateChecksums.Checksum)
                         continue;
-                    }
 
-                    if (oldProjectChecksums.ProjectReferences.Checksum != newProjectChecksums.ProjectReferences.Checksum)
-                    {
-                        solution = solution.WithProjectReferences(projectId, SpecializedCollections.EmptyEnumerable<ProjectReference>());
-                    }
+                    changedProjects.Add((projectId, oldProjectStateChecksums, newProjectStateChecksums));
                 }
 
-                // removed project
-                foreach (var (projectId, _) in oldMap)
+                // From this point on, we can change the solution instance.
+
+                // Second, add any projects on the 'new' side that aren't currently in this solution.
+                foreach (var (newProjectChecksum, newProjectId) in newSolutionChecksums.Projects)
                 {
-                    if (!newMap.ContainsKey(projectId))
+                    if (!oldProjectIds.Object.Contains(newProjectId))
                     {
-                        // we have a project removed
-                        solution = solution.RemoveProject(projectId);
+                        var projectInfo = await _assetProvider.CreateProjectInfoAsync(newProjectId, newProjectChecksum, cancellationToken).ConfigureAwait(false);
+                        Contract.ThrowIfNull(projectInfo);
+                        solution = solution.AddProject(projectInfo);
```

Comment:
>solution = solution.AddProject(projectInfo); [](http://example.com/codeflow?start=24&length=44)

I only ask because I'm now sensitive about all the forking going on, but would it make sense for there to be an operation that could add/remove/update multiple projects on the solution at once to reduce forking? Or is this code path almost always just causing a single fork?

Reply:
@ToddGrun  i think there very much should either:

1. be a helper that can do multiple at once
2. make forking cheaper, by deferring some of the work it does till much later.

tagging @jasonmalinowski 

>  Or is this code path almost always just causing a single fork?

This actually *rarely* happens after first sync.  It basically will only happen when projects get added (a rare case to begin with).

Where we need to care more is things like a linked doc changing, where the linked doc is in many TFMs.  this will happen constantly as users type.  And needs to be super fast.  That's the forking case we keep hitting, with our linked files that are in like a dozen tfms.

Reply:
Note that adding a new project here may create a fork in the sense it creates a new Solution + Solution[Compilation]State but by itself that's not the biggest deal since those are cheap objects. The real cost would be in the recreation of CompilationTrackers/project dependency graphs and such -- but since these are one-shot they're going to be about as cheap as they could get.

That said I don't imagine it'd be terribly hard to change if we wanted to, but I'd definitely want to see some good evidence we have a problem here and that problem is bigger than the other problems we could fix with the same engineering effort.